### ISSUE TITLE
Fine-tuning or training from scratch in a differente language?

### ISSUE THREAD

**Issue by paulovasconcellos-hotmart** [yl4579/StyleTTS2#197]
Hi everyone,
I'm considering putting some effort into training StyleTTS in Portuguese. I have a good-quality dataset for this task, however, I was in doubt if it would be better just to fine-tune the model (which I know was trained in English), or (since it's an unseen language) train the model from scratch in Portuguese.

Does anyone have some tips on what I should consider before making a decision?

**Comment by martinambrus** [comment]
Definitelly train a new PLBERT for a new language. You can try with the one trained in English but even [the author says](https://github.com/yl4579/StyleTTS2?tab=readme-ov-file#pre-trained-modules) it probably won't work.

**Comment by rlenain** [comment]
Hi there -- I have trained a PL-BERT model on a 14 language dataset which was crowdsourced by the author of the paper. You can find this model open-sourced here: https://huggingface.co/papercup-ai/multilingual-pl-bert

Using this PL-BERT model, you can now train multilingual StyleTTS2 models. In my experiments, I have found that you don't need to train from scratch in order to train multilingual StyleTTS2, you can just finetune. Follow the steps outlined in the link I shared above!

Best of luck, and let me know what you make with this!

**Comment by paulovasconcellos-hotmart** [comment]
Thank you very much for this @rlenain . Do you have any good resources on how to train StyleTTS2?

**Comment by Stardust-minus** [comment]
> Hi there -- I have trained a PL-BERT model on a 14 language dataset which was crowdsourced by the author of the paper. You can find this model open-sourced here: https://huggingface.co/papercup-ai/multilingual-pl-bert
> 
> Using this PL-BERT model, you can now train multilingual StyleTTS2 models. In my experiments, I have found that you don't need to train from scratch in order to train multilingual StyleTTS2, you can just finetune. Follow the steps outlined in the link I shared above!
> 
> Best of luck, and let me know what you make with this!

Nice work！Did the Chinese data the model used for training include tone？

**Comment by rlenain** [comment]
I'm not sure -- you can see a sample here (the data is from this dataset: https://huggingface.co/datasets/styletts2-community/multilingual-phonemes-10k-alpha/viewer/zh).
<img width="1512" alt="image" src="https://github.com/yl4579/StyleTTS2/assets/22459503/557d6ad7-bc3c-43dc-9d3f-fec4de65ff24">

**Comment by Frederieke93** [comment]
Thank you very much @rlenain! This is a great addition! You mentioned you can just finetune on a new language instead of training a new base model, I'd like to try it. How large are the datasets you used for the finetuning on a new language?

**Comment by rlenain** [comment]
i tend to keep some english in the dataset (~5 hours) and have had success with as little as 20 hours of Spanish data split across 4 speakers

**Comment by rlenain** [comment]
https://huggingface.co/papercup-ai/multilingual-pl-bert

**Comment by casic** [comment]
Thanks

На ср, 6.03.2024 г. в 13:06 Raphael Lenain ***@***.***>
написа:

**Comment by ZYJGO** [comment]
@rlenain > i tend to keep some english in the dataset (~5 hours) and have had success with as little as 20 hours of Spanish data split across 4 speakers

thanks for the great work! do you have some samples to share? I'm very curious about the quality on a new language

**Comment by rlenain** [comment]
Unfortunately because of the privacy policy of the samples that I trained on, I cannot share these samples here. What I can say is that the quality is very much on-par with samples you can find on the samples page in English.

**Comment by traderpedroso** [comment]
> Unfortunately because of the privacy policy of the samples that I trained on, I cannot share these samples here. What I can say is that the quality is very much on-par with samples you can find on the samples page in English.

 I would like to ask three questions: Do the speakers in the dataset need to be in a numeric format, for example, speaker 0, 1, 2, and do they have to be different from 0, or can I put all of them with the same name or even in a string format like a name to facilitate the recognition of the speakers? The other question is, after training the speakers, to access them, do I need to define the speakers in the inference and what about the language selector is automatic?

**Comment by ara-vardanyan** [comment]
> @rlenain > i tend to keep some english in the dataset (~5 hours) and have had success with as little as 20 hours of Spanish data split across 4 speakers
> 
> thanks for the great work! do you have some samples to share? I'm very curious about the quality on a new language

Hey! Is your Spanish fine-tune open source by chance?

**Comment by sch0ngut** [comment]
@rlenain 

> i tend to keep some english in the dataset (~5 hours) and have had success with as little as 20 hours of Spanish data split across 4 speakers

@rlenain Do you mind sharing for how many epochs you fine-tuned?

**Comment by rlenain** [comment]
@sch0ngut Generally for 50k-100k iterations, whatever that means in terms of epochs for the size of your dataset. But you should be following the validation curve.

**Comment by 21sK1p** [comment]
> Hi there -- I have trained a PL-BERT model on a 14 language dataset which was crowdsourced by the author of the paper. You can find this model open-sourced here: https://huggingface.co/papercup-ai/multilingual-pl-bert
> 
> Using this PL-BERT model, you can now train multilingual StyleTTS2 models. In my experiments, I have found that you don't need to train from scratch in order to train multilingual StyleTTS2, you can just finetune. Follow the steps outlined in the link I shared above!
> 
> Best of luck, and let me know what you make with this!



> Hi there -- I have trained a PL-BERT model on a 14 language dataset which was crowdsourced by the author of the paper. You can find this model open-sourced here: https://huggingface.co/papercup-ai/multilingual-pl-bert
> 
> Using this PL-BERT model, you can now train multilingual StyleTTS2 models. In my experiments, I have found that you don't need to train from scratch in order to train multilingual StyleTTS2, you can just finetune. Follow the steps outlined in the link I shared above!
> 
> Best of luck, and let me know what you make with this!

@rlenain  what would I need to do if have to train it in hindi language?

**Comment by rlenain** [comment]
You can probably just finetune StyleTTS2 without changing the PL-BERT model, and it would work, with the right data and amount of data. 
If you want to train PL-BERT on Hindi, I believe there's data here: https://huggingface.co/datasets/styletts2-community/multilingual-pl-bert

**Comment by JingchengYang4** [comment]
@rlenain Regarding this multilingual pl-bert, it appears the data used to train this model uses a data-processing script that's unavailable to the general public - how would we be able to tokenize the training data for StyleTTS in the same form as the Bert model?

**Comment by rlenain** [comment]
the data here (https://huggingface.co/datasets/styletts2-community/multilingual-pl-bert) has been tokenized using the tokenizer of the `bert-multilingual-base-cased` model: https://huggingface.co/google-bert/bert-base-multilingual-cased

**Comment by chocolatedesue** [comment]
Hello @rlenain,

I've successfully trained StyleTTS2 with the multilingual PL-BERT from this source during the first stage using the LJSpeech dataset provided in this repository.

However, I encountered an issue at the start of the second stage where NaN values appeared. Could you help me identify any potential mistakes?

Here's what I've done so far:
1. Converted the source WAV files to a 24k WAV format.
2. Replaced the files in Utils/PLBERT/ with the multilingual PL-BERT.
3. Conducted training on eight 3090 cards for 12 hours without any other modifications.

first stage loss graph
![image](https://github.com/yl4579/StyleTTS2/assets/91049859/d1081739-cc37-4974-9a86-f851056623b2)


Appended

1. with debug , i find the first nan comes from https://github.com/yl4579/StyleTTS2/blob/5cedc71c333f8d8b8551ca59378bdcc7af4c9529/train_second.py#L400

**Comment by chocolatedesue** [comment]
> Hello @rlenain,
> 
> I've successfully trained StyleTTS2 with the multilingual PL-BERT from this source during the first stage using the LJSpeech dataset provided in this repository.
> 
> However, I encountered an issue at the start of the second stage where NaN values appeared. Could you help me identify any potential mistakes?
> 
> Here's what I've done so far:
> 
> 1. Converted the source WAV files to a 24k WAV format.
> 2. Replaced the files in Utils/PLBERT/ with the multilingual PL-BERT.
> 3. Conducted training on eight 3090 cards for 12 hours without any other modifications.
> 
> first stage loss graph ![image](https://private-user-images.githubusercontent.com/91049859/328726898-d1081739-cc37-4974-9a86-f851056623b2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUxMzY1MDcsIm5iZiI6MTcxNTEzNjIwNywicGF0aCI6Ii85MTA0OTg1OS8zMjg3MjY4OTgtZDEwODE3MzktY2MzNy00OTc0LTlhODYtZjg1MTA1NjYyM2IyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA4VDAyNDMyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUwMGQ3ODM3OGMxMDE1ZjU3MWYyMTYyNzliOTc4YTA0MWNjNjIxMTlhYWZiMGJhODEzODZkNTNlYzQ4MTI5YjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.VAEse1NtGQataBV9gZ7Tn_vvA3DX7ud0gOdvBHx6Ngw)
> 
> Appended
> 
> 1. with debug , i find the first nan comes from https://github.com/yl4579/StyleTTS2/blob/5cedc71c333f8d8b8551ca59378bdcc7af4c9529/train_second.py#L400

solve it, just a bad config that casuing the first stage params loads to second stage model params

I should config first_stage_path instead of pretrained_model

**Comment by tanishbajaj101** [comment]
> Hello @rlenain,
> 
> I've successfully trained StyleTTS2 with the multilingual PL-BERT from this source during the first stage using the LJSpeech dataset provided in this repository.
> 
> However, I encountered an issue at the start of the second stage where NaN values appeared. Could you help me identify any potential mistakes?
> 
> Here's what I've done so far:
> 
> 1. Converted the source WAV files to a 24k WAV format.
> 2. Replaced the files in Utils/PLBERT/ with the multilingual PL-BERT.
> 3. Conducted training on eight 3090 cards for 12 hours without any other modifications.
> 
> first stage loss graph ![image](https://private-user-images.githubusercontent.com/91049859/328726898-d1081739-cc37-4974-9a86-f851056623b2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzNzc1MjUsIm5iZiI6MTcyMTM3NzIyNSwicGF0aCI6Ii85MTA0OTg1OS8zMjg3MjY4OTgtZDEwODE3MzktY2MzNy00OTc0LTlhODYtZjg1MTA1NjYyM2IyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE5VDA4MjAyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE4MWU5NGQ3MzFjNWY2NDViYjFiZWIyNWY4ZmRlNzI3NTM2ODFmZDIxNDQ4YTg1NTFhYjAzZmYxNGI0ZmE0ODEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.SN9c_5hK4GcuLEPUFPl8xLe96JiIAP0sJ_KkpE2X6_w)
> 
> Appended
> 
> 1. with debug , i find the first nan comes from https://github.com/yl4579/StyleTTS2/blob/5cedc71c333f8d8b8551ca59378bdcc7af4c9529/train_second.py#L400

can you please elaborate which files you replaced in PLBERT folder

**Comment by chocolatedesue** [comment]
> > Hello @rlenain,
> > I've successfully trained StyleTTS2 with the multilingual PL-BERT from this source during the first stage using the LJSpeech dataset provided in this repository.
> > However, I encountered an issue at the start of the second stage where NaN values appeared. Could you help me identify any potential mistakes?
> > Here's what I've done so far:
> > 
> > 1. Converted the source WAV files to a 24k WAV format.
> > 2. Replaced the files in Utils/PLBERT/ with the multilingual PL-BERT.
> > 3. Conducted training on eight 3090 cards for 12 hours without any other modifications.
> > 
> > first stage loss graph ![image](https://private-user-images.githubusercontent.com/91049859/328726898-d1081739-cc37-4974-9a86-f851056623b2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzNzc1MjUsIm5iZiI6MTcyMTM3NzIyNSwicGF0aCI6Ii85MTA0OTg1OS8zMjg3MjY4OTgtZDEwODE3MzktY2MzNy00OTc0LTlhODYtZjg1MTA1NjYyM2IyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE5VDA4MjAyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE4MWU5NGQ3MzFjNWY2NDViYjFiZWIyNWY4ZmRlNzI3NTM2ODFmZDIxNDQ4YTg1NTFhYjAzZmYxNGI0ZmE0ODEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.SN9c_5hK4GcuLEPUFPl8xLe96JiIAP0sJ_KkpE2X6_w)
> > Appended
> > 
> > 1. with debug , i find the first nan comes from https://github.com/yl4579/StyleTTS2/blob/5cedc71c333f8d8b8551ca59378bdcc7af4c9529/train_second.py#L400
> 
> can you please elaborate which files you replaced in PLBERT folder


just cp the multilingual pl-bert to the old folder, it works perfectly

**Comment by Geremia** [comment]
@rlenain Did you use [OpenSLR](https://www.openslr.org/94/)'s recordings for Spanish, or something else?

**Comment by martinambrus** [comment]
@chocolatedesue would you mind closing the issue if it's solved now?

**Comment by chocolatedesue** [comment]
> @chocolatedesue would you mind closing the issue if it's solved now?

I think you might have mistaken me for someone else. 

This issue was not opened by me. It might be better to reach out directly to the original issuer, @paulovasconcellos-hotmart.

**Comment by martinambrus** [comment]
> > @chocolatedesue would you mind closing the issue if it's solved now?
> 
> I think you might have mistaken me for someone else.
> 
> This issue was not opened by me. It might be better to reach out directly to the original issuer, @paulovasconcellos-hotmart.

Sorry, I probably clicked the wrong name in autocomplete and didn't realize it. Thanks for pinging the original author :)

**Comment by jannikstdl** [comment]
Ok now we got the multilingual PLBERT Model, any ressources on how to finetune StyleTTS2 on a new language?

**Comment by jerryola1** [comment]
> Ok now we got the multilingual PLBERT Model, any ressources on how to finetune StyleTTS2 on a new language?

I think someone already answered this in the previous response

**Comment by thewh1teagle** [comment]
> Hi there -- I have trained a PL-BERT model on a 14 language dataset which was crowdsourced by the author of the paper. You can find this model open-sourced here: https://huggingface.co/papercup-ai/multilingual-pl-bert
> 
> Using this PL-BERT model, you can now train multilingual StyleTTS2 models. In my experiments, I have found that you don't need to train from scratch in order to train multilingual StyleTTS2, you can just finetune. Follow the steps outlined in the link I shared above!
> 
> Best of luck, and let me know what you make with this!

hi! i'm trying to fine-tune PL-BERT for Hebrew.
i've got a dataset of ~5M sentences in the format: `text<TAB>phonemes`.
all the phonemes exist in english, but since Hebrew has diacritics, I think it should be tokenized at the character level instead of word level.

do I need to change anything specific in the tokenizer or model config to make this work?
any tips for training PL-BERT with this kind of data would be super helpful. thanks!
